Add functions and a command for reorder rules in a policy file

fr0m-scratch opened the pull request:
1. Add a function ReorderDelegations() to handle rule reordering from backend
2. Add a command ```gittuf policy reorder-rules [flags]``` to reorder rules from CLI

fr0m-scratch commented:
> This looks good! A couple of items:
> 
> * Each commit to be merged into gittuf needs a DCO signoff, which can be achieved with the `-s` flag when running `git commit` [(see here)](https://wiki.linuxfoundation.org/dco) . You'll need to amend the commit and then force-push.
> * `ReorderDelegations` in `repository/targets.go` and `policy/targets.go` will need tests. See [here](https://github.com/gittuf/gittuf/blob/main/internal/repository/targets_test.go#L132) and [here](https://github.com/gittuf/gittuf/blob/main/internal/policy/targets_test.go#L48) for examples. You can copy part of the existing tests and just add code that first reorders and then checks that reordering was successful.
> * I'm unsure on the ordering format at the moment (i.e. comma separated or otherwise), I need to look a bit more into that.

Thank you for your comments! I've made the requested change and tested out. Please help me check it : )

neilnaveen commented on internal/policy/targets.go:
```diff
@@ -96,6 +96,27 @@ func UpdateDelegation(targetsMetadata *tuf.TargetsMetadata, ruleName string, aut
+			return nil, ErrCannotManipulateAllowRule
+		}
+
+		for _, delegation := range targetsMetadata.Delegations.Roles {
+			if delegation.Name == ruleName {
+				allDelegations = append(allDelegations, delegation)
+			}
+		}
```
Instead of doing this for every single rulename, can we first precompute this in a map of rulename to roles? Then, when we iterate through the rule names, we can query the roles for that rule name and append it to allDelegations instead of iterating multiple times.

fr0m-scratch replied:
Thanks, I'll try that with a pre-computed map

neilnaveen commented on internal/cmd/policy/reorderrules/reorderrules.go:
```diff
@@ -0,0 +1,69 @@
+type options struct {
+	p          *persistent.Options
+	policyName string
+	ruleNames  string
```
Instead of using a string, we could use cobra's StringArray, to take in an array of string, which means if the user mistakenly adds whitespace in front of a rulename, the reorder will still work

fr0m-scratch replied:
Thanks, great idea!

neilnaveen reviewed (commented):
This looks good, a couple of suggestions.

fr0m-scratch commented on internal/policy/targets_test.go:
```diff
@@ -86,6 +86,39 @@ func TestUpdateDelegation(t *testing.T) {
+	}
+
+	assert.Contains(t, targetsMetadata.Delegations.Keys, key1.KeyID)
+	assert.Contains(t, targetsMetadata.Delegations.Keys, key2.KeyID)
```
Thanks, copy that

fr0m-scratch commented on internal/cmd/policy/reorderrules/reorderrules.go:
```diff
@@ -0,0 +1,69 @@
+	cmd := &cobra.Command{
+		Use:   "reorder-rules",
+		Short: "Reorder rules in a policy file",
+	}
```
Thanks, gocha

fr0m-scratch commented:
> @fr0m-scratch Can you squash commits?

Sure, copy that

adityasaky commented:
I lean towards error-ing out. Too easy to forget a rule in the CLI. That said, can you record this flow using asciinema or something to get an idea of how this feature is being implemented? thanks!

fr0m-scratch commented:
I added checks for missing rules as well as relevant tests and squashed to one commit. Below is the link for my asciinema demo recording: https://asciinema.org/a/LnFiudjQD1RmgsnYb1Uy5b0TH

fr0m-scratch commented:
> Another edge case which we will want to catch: duplicate rule names with the correct number of rules total. For example, if the policy has `r1, r2, r3` as rules, calling `reorder-rules --rule-order r1 r2 r2` will work.
> 
> There's a `set` package that should be useful for this PR: https://github.com/gittuf/gittuf/blob/main/internal/common/set/set.go
> 
> Let's try and see how we can use that to simplify some operations in the PR (e.g. get the `Contents` of one set and loop over them, calling `Has` on the other set)

Appreciate for helping find the edge cases! Resolved with the suggested approach.
<img width="1342" alt="Screenshot 2024-08-02 at 2 56 48 AM" src="https://github.com/user-attachments/assets/5f35a4ed-c643-455c-89e7-15943e42d708">

fr0m-scratch commented on internal/repository/targets_test.go:
```diff
@@ -167,6 +167,74 @@ func TestUpdateDelegation(t *testing.T) {
+		assert.Equal(t, expectedFinalOrder, finalOrder)
+	})
+
+	t.Run("invalid rule names", func(t *testing.T) {
```
Copy that! I've pushed a new commit; please let me know If assert.ErrorIs() is good enough or I should make the asserts more specific.

adityasaky commented on internal/cmd/policy/policy.go:
```diff
@@ -34,6 +35,7 @@ func New() *cobra.Command {
 	cmd.AddCommand(removerule.New(o))
 	cmd.AddCommand(sign.New(o))
 	cmd.AddCommand(updaterule.New(o))
+	cmd.AddCommand(reorderrules.New(o))
```
nit: let's order this alphabetically so this goes after removerule.

adityasaky commented on internal/cmd/policy/reorderrules/reorderrules.go:
```diff
@@ -0,0 +1,70 @@
+
+	cmd.Flags().StringSliceVar(
+		&o.ruleNames,
+		"rule-order",
```
```suggestion
		"rule-name",
```

adityasaky commented on internal/cmd/policy/reorderrules/reorderrules.go:
```diff
@@ -0,0 +1,70 @@
+		&o.ruleNames,
+		"rule-order",
+		[]string{},
+		"a space-separated list of rule names, in the new order that they should appear in, from first to last",
```
does the flag not need to be repeated? as in `--rule-name <> --rule-name <> ...`

fr0m-scratch replied:
It can work both ways, either repeated or not repeated

adityasaky commented on internal/cmd/policy/reorderrules/reorderrules.go:
```diff
@@ -0,0 +1,70 @@
+		return err
+	}
+
+	err = repo.ReorderDelegations(cmd.Context(), signer, o.policyName, o.ruleNames, true)
```
just return here?

adityasaky commented on internal/cmd/policy/reorderrules/reorderrules.go:
```diff
@@ -0,0 +1,70 @@
+	cmd := &cobra.Command{
+		Use:               "reorder-rules",
+		Short:             "Reorder rules in the specified policy file",
+		Long:              `This command allows users to reorder rules in the specified policy file. Specify the names of rules in the new order they should appear in, starting from first to last rule. By default, the main policy file is selected. Note that authorized keys can be specified from disk, from the GPG keyring using the "gpg:<fingerprint>" format, or as a Sigstore identity as "fulcio:<identi ...
```
nit: I'd write this more like "The rule names must be specified in the new order they must appear in...".

We also don't have authorized keys here, no?

adityasaky commented on internal/policy/targets.go:
```diff
@@ -96,6 +102,54 @@ func UpdateDelegation(targetsMetadata *tuf.TargetsMetadata, ruleName string, aut
+	ruleSet := set.NewSet[string]()
+
+	for _, ruleName := range ruleNames {
+		if ruleName == AllowRuleName {
```
this brings up a good question: should users be specifying the allow-rule? IMO it's a convenience for us, so they should specify everything but.

fr0m-scratch replied:
Indeed, I think the allow-rule is more like a default setting which sits below user-level abstraction. And `policy --list-rules` does not include the allow-rule either.

adityasaky commented on internal/policy/targets.go:
```diff
@@ -96,6 +102,54 @@ func UpdateDelegation(targetsMetadata *tuf.TargetsMetadata, ruleName string, aut
+			return nil, fmt.Errorf("%w: %s", ErrDuplicatedRuleName, ruleName)
+		}
+
+		if delegation, valid := rolesMap[ruleName]; valid {
```
s/valid/exists -> we're not checking validity but that we know the delegation exists.

adityasaky commented on internal/policy/targets.go:
```diff
@@ -96,6 +102,54 @@ func UpdateDelegation(targetsMetadata *tuf.TargetsMetadata, ruleName string, aut
+	}
+
+	if len(missingRules) > 0 {
+		return nil, fmt.Errorf("%w: %s", ErrMissingRules, missingRules)
```
I'd strings.Join the missingRules with a `, `, a comma and space basically.

adityasaky commented on internal/policy/targets.go:
```diff
@@ -96,6 +102,54 @@ func UpdateDelegation(targetsMetadata *tuf.TargetsMetadata, ruleName string, aut
 }
 
+// ReorderDelegations changes the order of delegations in TargetsMetadata.
+func ReorderDelegations(targetsMetadata *tuf.TargetsMetadata, ruleNames []string) (*tuf.TargetsMetadata, error) {
```
I think this could generally be simplified a bit... cc @patzielinski

fr0m-scratch replied:
I haven't figure out a way to do this, but all other requested changes have been made.

adityasaky replied:
Taking a stab at making this easier to read:


```
func ReorderDelegations(targetsMetadata *tuf.TargetsMetadata, ruleNames []string) (*tuf.TargetsMetadata, error) {
	// Create a map of all existing delegations for quick look up
	rolesMap := make(map[string]tuf.Delegation)
	// Create a set of current rules in metadata, skipping the allow rule
	currentRules := set.NewSet[string]()
	for _, delegation := range targetsMetadata.Delegations.Roles {
		if delegation.Name == AllowRuleName {
			continue
		}
		rolesMap[delegation.Name] = delegation
		currentRules.Add(delegation.Name)
	}

	specifiedRules := set.NewSet[string]()
	for _, name := range ruleNames {
		if specifiedRules.Has(name) {
			return nil, fmt.Errorf("%w: '%s'", ErrDuplicatedRuleName, name)
		}
		specifiedRules.Add(name)
	}

	onlyInSpecifiedRules := specifiedRules.Minus(currentRules)
	if onlyInSpecifiedRules.Len() != 0 {
		if onlyInSpecifiedRules.Has(AllowRuleName) {
			return nil, fmt.Errorf("%w: do not specify allow rule", ErrCannotManipulateAllowRule)
		}

		contents := onlyInSpecifiedRules.Contents()
		return nil, fmt.Errorf("%w: rules '%s' do not exist in current rule file", ErrRuleNotFound, strings.Join(contents, ", "))
	}

	onlyInCurrentRules := currentRules.Minus(specifiedRules)
	if onlyInCurrentRules.Len() != 0 {
		contents := onlyInCurrentRules.Contents()
		return nil, fmt.Errorf("%w: rules '%s' not specified", ErrRuleNotFound, strings.Join(contents, ", "))
	}

	// Create newDelegations and set it in the targetsMetadata after adding allow rule
}
```

I think this may be slightly less optimal as we will end up iterating over `ruleNames` at least twice directly and some more in the implementation of `set.Minus` (which I haven't shown here), but I think that's okay because:
a) reordering is unlikely to be a regular operation
b) some of the cost (eg. `set.Contents()`) is going to only be during error cases

I personally think the result ends up being more readable.

adityasaky replied:
We could also only do the `.Minus` when `specifiedRules` and `currentRules` are not equal, as `.Minus` is really only needed for the error message to be helpful.

fr0m-scratch replied:
I haven't found the implementation for `.Minus`, so I added a quick impl in the common/set package

adityasaky commented on internal/repository/targets.go:
```diff
@@ -200,6 +200,66 @@ func (r *Repository) UpdateDelegation(ctx context.Context, signer sslibdsse.Sign
+	}
+
+	slog.Debug("Checking if all rules exist...")
+	for _, ruleName := range ruleNames {
```
let's not split checks between here and the policy package, a future refactor could accidentally include one but not the other, etc.

adityasaky reviewed (commented):
This is pretty close!

adityasaky commented on internal/policy/targets_test.go:
```diff
@@ -86,6 +86,53 @@ func TestUpdateDelegation(t *testing.T) {
+		t.Fatal(err)
+	}
+
+	assert.Contains(t, targetsMetadata.Delegations.Keys, key1.KeyID)
```
I appreciate the thoroughness here :smile: 


That said, I think we can remove all of these assert statements as we want to (and do) test that in `TestAddDelegation`.

adityasaky commented on internal/policy/targets_test.go:
```diff
@@ -86,6 +86,53 @@ func TestUpdateDelegation(t *testing.T) {
+	ruleNames := []string{"rule-2", "rule-1"}
+	targetsMetadata, err = ReorderDelegations(targetsMetadata, ruleNames)
+	assert.Nil(t, err)
+
```
let's set up a table of tests here:

```
	tests := map[string]struct {
		newOrder      []string
		expectedError error
	}{
		"reverse order": {...},
		"rule not specified in new order": {...},
		"rule repeated in new order": {...},
		"unknown rule in new order": {...},
		"allow rule in new order": {...},
	}
```

I think it's also easier for us to be sure we're testing every case, which I think isn't happening right now?

adityasaky replied:
See: https://github.com/gittuf/gittuf/blob/bb9ced2c12c8ec070e59c6f9d600ebc9a22197b1/internal/gitinterface/hash_test.go#L15-L61

fr0m-scratch replied:
Will do!

adityasaky commented on internal/repository/targets_test.go:
```diff
@@ -167,6 +167,87 @@ func TestUpdateDelegation(t *testing.T) {
+		assert.Equal(t, expectedInitialOrder, initialOrder)
+
+		// Case 1: Valid Input
+		newOrder := []string{"rule-3", "rule-1", "rule-2", "protect-main"}
```
Ah I see, the tests are more complete here. I think it makes more sense to test in the policy package, which is where we actually catch the different errors.

fr0m-scratch replied:
It does make sense! Now, I kept only the valid case and move the complete test suite to the policy package

adityasaky commented on internal/cmd/policy/reorderrules/reorderrules.go:
```diff
@@ -0,0 +1,66 @@
+}
+
+func (o *options) Run(cmd *cobra.Command, args []string) error {
+	o.ruleNames = append(o.ruleNames, args...)
```
I missed this line before: what's this for?

fr0m-scratch replied:
Yes, it ensures `--rule-name` works either way

adityasaky replied:
hm, is there anything that stops a rule name from having a space in it? This would break those cases (compared to `--rule-name "rule 1" --rule-name "rule 2"` I think.

cc @patzielinski

fr0m-scratch replied:
If we use multiple quotes following `rule-name`, it can still parse those with space in them. Though we probably should specify that to users in the doc

<img width="1338" alt="Screenshot 2024-09-06 at 4 45 34 PM" src="https://github.com/user-attachments/assets/2a1add5b-0ff5-4c38-b0a0-31fef9cb6f88">
<img width="1346" alt="Screenshot 2024-09-06 at 4 45 52 PM" src="https://github.com/user-attachments/assets/7a900582-1c85-4b71-a10c-3be0b9d49711">

adityasaky replied:
thinking about this some more, do we even need `--rule-name` then? Can we just use `args`?

fr0m-scratch replied:
Do you mean the rule names goes directly as `args`? I suppose we can still keep `--rule-name`, just to signal it is the start of specifying the new order

adityasaky replied:
No, I think we could just make that clear in the docs for the subcommand. Having a flag can be confusing if we really want them to pass in args. Let's get rid of the flag and use args only, and make that clear in the command docs, thanks!

fr0m-scratch replied:
Gocha, I'll remove the flag and update the docs

adityasaky commented on internal/policy/targets.go:
```diff
@@ -96,6 +103,58 @@ func UpdateDelegation(targetsMetadata *tuf.TargetsMetadata, ruleName string, aut
+		specifiedRules.Add(name)
+	}
+
+	if currentRules.Len() != specifiedRules.Len() {
```
This check is insufficient. If currentRules is `rule-1, rule2` and specifiedRules is `rule-1, rule-3`, we should error. We explicitly need a `set.Equal()` check (which in turn _can_ use .Len() as a first heuristic to return false as two sets cannot be equal if they're of different lengths).

fr0m-scratch replied:
Will do!

adityasaky commented on internal/policy/targets.go:
```diff
@@ -96,6 +103,58 @@ func UpdateDelegation(targetsMetadata *tuf.TargetsMetadata, ruleName string, aut
+	}
+
+	// Create newDelegations and set it in the targetsMetadata after adding allow rule
+	newDelegations := []tuf.Delegation{}
```
```suggestion
	newDelegations := make([]tuf.Delegation, 0, len(rolesMap) + 1)
```

adityasaky commented on internal/policy/targets_test.go:
```diff
@@ -86,6 +86,75 @@ func TestUpdateDelegation(t *testing.T) {
+			ruleNames:     []string{"rule-3", "rule-2", "rule-1", "rule-3"},
+			expectedError: ErrDuplicatedRuleName,
+		},
+		"unknown rule in the new order": {
```
as noted above, this should also check when the specified list is the right length. Without this, we can panic when looking up a rule in rolesMap.

adityasaky reviewed (approved):
Some minor nits, but I think this is otherwise good to go. Great work!

fr0m-scratch commented:
> Some minor nits, but I think this is otherwise good to go. Great work!

Have committed the new changes, appreciate it!

adityasaky reviewed (approved):
This looks great, thanks for working on this!

adityasaky commented on internal/cmd/policy/reorderrules/reorderrules.go:
```diff
@@ -0,0 +1,58 @@
+	cmd := &cobra.Command{
+		Use:               "reorder-rules",
+		Short:             "Reorder rules in the specified policy file",
+		Long:              "This command allows users to reorder rules in the specified policy file. By default, the main policy file is selected. The rule names need to be passed as arguments, in the new order they must appear in, starting from the first to the last rule. Rule names may contain spaces, so they should be enclosed in quotes if necessary.",
```
No, the user isn't really aware of it at all. They can only encounter it when they go looking at the raw policy metadata, it's also not part of `policy list-rules`.
